formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufOneOfTest.kt

```diff
@@ -0,0 +1,464 @@
+            /**
+             * 7: {5: 32}
+             * 3: {"foo"}
+             */
+            assertEquals("3a0228201a03666f6f", it)
+        }
+    }
+
+    @Test
+    fun testDecodeNestedStruct() {
+        val data = OneOfData(NestedIntType(IntType(32)), "foo")
+        ProtoBuf.decodeFromHexString<OneOfData>("3a0228201a03666f6f").also {
+            println(it)
+            assertEquals(data, it)
+        }
+    }
+
+    @Serializable
+    @ProtoNumber(8)
+    data class FailType(val i: Int, val j: Int) : IType
+
+    @Test
+    fun testOneOfElementCheck() {
+        val data = OneOfData(FailType(1, 2), "foo")
+        assertFailsWith<IllegalArgumentException> { ProtoBuf.encodeToHexString(OneOfData.serializer(), data) }
+    }
+
+    @Serializable
+    @ProtoNumber(9)
+    data class NestedOneOfType(val i: InnerNested) : IType
+
+    @Serializable
+    data class InnerNested(@ProtoOneOf val i: InnerOneOf)
+
+    @Serializable
+    sealed interface InnerOneOf
+
+    @Serializable
+    @ProtoNumber(1)
+    data class InnerInt(val i: Int) : InnerOneOf
+
+    @Serializable
+    @ProtoNumber(2)
+    data class InnerString(val s: String) : InnerOneOf
+
+    @Test
+    fun testEncodeNestedOneOf() {
+        val data = OneOfData(NestedOneOfType(i = InnerNested(InnerInt(32))), "foo")
+        ProtoBuf.encodeToHexString(OneOfData.serializer(), data).also {
+            println(it)
+            /**
+             * 9: {1: 32}
+             * 3: {"foo"}
+             */
+            assertEquals("4a0208201a03666f6f", it)
+        }
+    }
+
+    @Test
+    fun testDecodeNestedOneOf() {
+        val data = OneOfData(NestedOneOfType(i = InnerNested(InnerInt(32))), "foo")
+        ProtoBuf.decodeFromHexString<OneOfData>("4a0208201a03666f6f").also {
+            println(it)
+            /**
+             * 9: {1: 32}
+             * 3: {"foo"}
+             */
+            assertEquals(data, it)
+        }
+    }
+
+    @Serializable
+    data class DoubleOneOfElement(
+        @ProtoOneOf val one: IType,
+        @ProtoNumber(3) val name: String,
+        @ProtoOneOf val two: OtherType
+    )
+
+    interface OtherType
+
+    @Serializable
+    @ProtoNumber(4)
+    data class OtherIntType(val i: Int) : OtherType
+
+    @Serializable
+    @ProtoNumber(5)
+    data class OtherStringType(val s: String) : OtherType
+
+    @Test
+    fun testEncodeDoubleOneOf() {
+        val module = SerializersModule {
+            polymorphic(OtherType::class) {
+                subclass(OtherStringType::class)
+            }
+        }
+        val buf = ProtoBuf {
+            serializersModule = module
+        }
+        val data = DoubleOneOfElement(
+            IntType(32),
+            "foo",
+            OtherStringType("bar")
+        )
+        buf.encodeToHexString(DoubleOneOfElement.serializer(), data).also {
+            println(it)
+            assertEquals("08201a03666f6f2a03626172", it)
+        }
+
+        assertFailsWith<SerializationException> {
```

Comment:
The only reason for failure here is that `OtherIntType` is not registered in the module, right?

Reply:
Yes, message check added here